### PR TITLE
513-Add leaderboard display mode

### PR DESCRIPTION
**Pull request by AuroraLauu**
# Template
This is a template. While not all sections are necessary, depending on the size and complexity of the PR,
more information is better. 

# @ mention of reviewers
@ckcollab @dde6khkg 
...

# A brief description of the purpose of the changes contained in this PR.
...

# Issues this PR resolves
...
https://github.com/codalab/competitions-v2/issues/513

# Known issues to be addressed in a separate PR
...

# A checklist for hand testing
- [ ] add checklist here

# Any relevant files for testing
[link]('#') to any relevant files (or drag and drop into github)

# Misc. comments
...

# Checklist
- [ ] Code review by me 
- [ ] Hand tested by me 
- [ ] I'm proud of my work
- [ ] Code review by reviewer
- [ ] Hand tested by reviewer
- [ ] Ready to merge

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -30,3 +30,7 @@ docker-compose.override.yml
 
 server_config.yaml
 /graphs/
+
+.idea/
```
Comment (ckcollab) [inline .gitignore:34]: these should exist in a global git ignore

File: src/apps/api/serializers/competitions.py [hunk 2]
```diff
@@ -123,6 +123,7 @@ class Meta:
             'docker_image',
             'allow_robot_submissions',
             'competition_type',
+            'leaderboard_mode',
```
Comment (ckcollab) [inline src/apps/api/serializers/competitions.py:126]: leaderboard mode should be attached to either phase or leaderboard, I think? some phases may want different flags

File: src/apps/competitions/models.py [hunk 3]
```diff
@@ -22,12 +22,23 @@
...
+    BEST = 'best'
 
     COMPETITION_TYPE = (
         (COMPETITION, "competition"),
         (BENCHMARK, "benchmark"),
     )
 
+    LEADERBOARD_MODE = (
+        (MANUALLY, 'manually'),
+        (LATSET, 'latest'),
```
Comment (ckcollab) [inline src/apps/competitions/models.py:37]: typo, `LATSET` -> `LATEST`

File: src/apps/api/serializers/competitions.py (hunk 2 above)
Comment (Logan-Ruf) [reply src/apps/api/serializers/competitions.py:126]: I recently added this to the `Leaderboard` model. I didn't realize you were currently working on this task, but we could probably move over to using these flags.
`
    ADD = "Add"
    ADD_DELETE = "Add_And_Delete"
    ADD_BY_REVIEW = "Add_By_Review"
    FORCE_LAST = "Force_Last"
    FORCE_LATEST_MULTIPLE = "Force_Latest_Multiple"
    FORCE_BEST = "Force_Best"
    SUBMISSION_RULES = (
        (ADD, "Only allow adding one submission"),
        (ADD_DELETE, "Allow users to add a single submission and remove that submission"),
        (ADD_BY_REVIEW, "Only allow adding one submission that is approved by review"),
        (FORCE_LAST, "Force only the last submission"),
        (FORCE_LATEST_MULTIPLE, "Force latest submission to be added to leaderboard (multiple)"),
        (FORCE_BEST, 'Force only the best submission to the leaderboard')
    )

`

File: src/apps/competitions/models.py (hunk 3 above)
Comment (AuroraLauu) [reply src/apps/competitions/models.py:37]: OMG，thanks for pointing out 👍

File: .gitignore (hunk 1 above)
Comment (AuroraLauu) [reply .gitignore:34]: okay

File: src/apps/api/serializers/competitions.py (hunk 2 above)
Comment (AuroraLauu) [reply src/apps/api/serializers/competitions.py:126]: @ckcollab okay, sounds great

Comment (AuroraLauu) [reply src/apps/api/serializers/competitions.py:126]: @Logan-Ruf okay, so what i am going to do is use these flags instead of current pr flags? 👍

**Comment by AuroraLauu** [comment]
@ckcollab @Logan-Ruf  i already update my pr, can you take a review on this, thanks~

File: src/apps/leaderboards/strategies.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        elif Leaderboard.FORCE_LATEST_MULTIPLE == submission_rule:
+            return AllModeStrategy()
+        elif Leaderboard.FORCE_BEST == submission_rule:
+            return BestModeStrategy()
+
+
+def put_on_leaderboard_by_submission_rule(request, submission_pk, submission_rule):
+    """add submission score to leaderboard by display strategy"""
+    strategy = StrategyFactory.create_by_submission_rule(submission_rule)
+    strategy.put_on_leaderboard(request, submission_pk)
```
Comment (Logan-Ruf) [inline src/apps/leaderboards/strategies.py:123]: Front-end needs to pull submission list again after submission is run now.

**Review by Logan-Ruf — CHANGES_REQUESTED** [review]
Do you want to add an input for changing the leaderboard rule on the competition edit form and add an optional field in the v2 YAML ingestion?

Should we keep the logic for `submission_leaderboard_connection` where it is at, or do you want to move that into `strategies.py`

File: src/apps/leaderboards/strategies.py (hunk 4 above)
Comment (AuroraLauu) [reply src/apps/leaderboards/strategies.py:123]: ok

**Review by AuroraLauu — COMMENTED** [review]
> Do you want to add an input for changing the leaderboard rule on the competition edit form and add an optional field in the v2 YAML ingestion?
> 
> Should we keep the logic for `submission_leaderboard_connection` where it is at, or do you want to move that into `strategies.py`

add an input for changing the leaderboard rule and add an optional field in the v2 YAML ingestion is good!
i think keep the logic for `submission_leaderboard_connection`  where it is at

File: src/apps/leaderboards/strategies.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+        self.update_submission(submission=submission,
+                               submission_pk=submission_pk,
+                               leaderboard=leaderboard)
+
+    def do_execute(self, phase, request, submission):
+        pass
+
+
+class ManuallyModeStrategy(BaseModeStrategy):
```
[the remark this answers is not in the document]
Comment (AuroraLauu) [reply src/apps/leaderboards/strategies.py]: sir, could you also change `ManuallyModeStrategy` to `ManualModeStrategy` in StrategyFactory？

File: src/apps/leaderboards/strategies.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    def __str__(self):
+        return "BestModeStrategy"
+
+
+class StrategyFactory:
+
+    @staticmethod
+    def create_by_submission_rule(submission_rule):
+        if submission_rule is None:
+            return ManuallyModeStrategy()
```
Comment (Logan-Ruf) [inline src/apps/leaderboards/strategies.py:111]:
```suggestion
            return ManualModeStrategy()
```
Changing `StrategyFactory` return for `ManualModeStrategy`

**Review by Logan-Ruf — APPROVED** [review]

**Comment by AuroraLauu** [comment]
@ckcollab @Logan-Ruf  can we merge this PR?

**Pull request merged**